sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java

```diff
@@ -322,8 +330,71 @@ public static void rename(
     if (srcToRename.isEmpty()) {
       return;
     }
-    getFileSystemInternal(srcToRename.iterator().next().getScheme())
-        .rename(srcToRename, destToRename);
+
+    boolean replaceExisting =
+        options.contains(MoveOptions.StandardMoveOptions.REPLACE_EXISTING) ? true : false;
+    rename(
+        getFileSystemInternal(srcToRename.iterator().next().getScheme()),
+        srcToRename,
+        destToRename,
+        replaceExisting);
+  }
+
+  /**
+   * Executes a rename of the src which all must exist using the provided filesystem.
+   *
+   * <p>If replaceExisting is enabled and filesystem throws {code FileAlreadyExistsException} then
+   * an attempt to delete the destination is made and the rename is retried. Some filesystem
+   * implementations may apply this automatically without throwing.
+   *
+   * @param fileSystem The filesystem in use
+   * @param srcResourceIds The source resources to move
+   * @param destResourceIds The destinations for the sources to move to (must be same length as
+   *     srcResourceIds)
+   * @param replaceExisting If existing files in destination should be overwritten
+   * @throws IOException If the rename could not be completed
+   */
+  @VisibleForTesting
+  static void rename(
+      FileSystem fileSystem,
+      List<ResourceId> srcResourceIds,
+      List<ResourceId> destResourceIds,
+      boolean replaceExisting)
+      throws IOException {
+    try {
+      fileSystem.rename(srcResourceIds, destResourceIds);
+    } catch (FileAlreadyExistsException e) {
```

Comment:
Thanks @chamikaramj 

> According to FileSystem.rename() we should be catching java.io.FileNotFoundException instead of java.nio.file.FileAlreadyExistsException.

I'm afraid this PR is all about the destination paths (FileAlreadyExistsException) and nothing to do with the source (FileNotFoundException).

FileSystem.rename() was in inadequately documented IMO as it only took into account files no longer existing in the sources (e.g. to accommodate failure/retry) and overlooked that implementations may also throw when the targets exist too. To my knowledge HDFS is the only one that does (see my recent commit on it where it now throws instead of swallowing and leaving corrupt data). 

I designed this so that any FS that threw a FileAlreadyExistsException could be handled and the caller could opt to delete and retry or not.

Given the confusion my implementation has given, I now believe it sensible to close this PR and I make HDFSFileSystem handle this internally. It will then always overwrite existing files (like other FS do). I didn't want to do that originally because it is not intuitive - Hadoop devs are used to jobs failing when targets exist. The reality is though, that we need force that anyway for the FileBasedSink.

@iemejia @chamikaramj  - would you be ok with that approach please? Thank you both for taking the time to be involved in this, and I am sorry that I have not been able to provide an intuitive solution - the questions it's raised convince me that it is a confusing patch so shouldn't be merged.